v1/status_message.go

```diff
@@ -0,0 +1,147 @@
+	ApplicationMetadataLayerSigPubKey *ecdsa.PublicKey
+}
+
+// SigPubKey returns the most important signature, from the application layer to transport
+func (s *StatusMessage) SigPubKey() *ecdsa.PublicKey {
+	if s.ApplicationMetadataLayerSigPubKey != nil {
+		return s.ApplicationMetadataLayerSigPubKey
+	}
+
+	return s.TransportLayerSigPubKey
+}
+
+func (s *StatusMessage) Clone() (*StatusMessage, error) {
+	copy := &StatusMessage{}
+
+	err := copier.Copy(&copy, s)
+	return copy, err
+}
+
+func (m *StatusMessage) HandleTransport(shhMessage *whisper.Message) error {
+	publicKey, err := crypto.UnmarshalPubkey(shhMessage.Sig)
+	if err != nil {
+		return errors.Wrap(err, "failed to get signature")
+	}
+
+	m.TransportMessage = shhMessage
+	m.Hash = shhMessage.Hash
+	m.TransportLayerSigPubKey = publicKey
+	m.TransportPayload = shhMessage.Payload
+
+	return nil
+}
+
+func (m *StatusMessage) HandleEncryption(myKey *ecdsa.PrivateKey, senderKey *ecdsa.PublicKey, enc *encryption.Protocol) error {
+	// As we handle non-encrypted messages, we make sure that DecryptPayload
+	// is set regardless of whether this step is successful
+	m.DecryptedPayload = m.TransportPayload
+
+	var protocolMessage encryption.ProtocolMessage
+	err := proto.Unmarshal(m.TransportPayload, &protocolMessage)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal ProtocolMessage")
+	}
+
+	payload, err := enc.HandleMessage(
+		myKey,
+		senderKey,
+		&protocolMessage,
+		m.Hash,
+	)
+
+	// For any error, we return it and assign
+	if err != nil {
+		return errors.Wrap(err, "failed to handle Encryption message")
+	}
+
+	m.DecryptedPayload = payload
+	return nil
+}
+
+func (m *StatusMessage) HandleDatasync(datasync *datasync.DataSync) ([]*StatusMessage, error) {
+	var statusMessages []*StatusMessage
+	payloads := datasync.Handle(
+		m.SigPubKey(),
+		m.DecryptedPayload,
+	)
+	for _, payload := range payloads {
+		message, err := m.Clone()
+		if err != nil {
+			return nil, err
+		}
+		message.DecryptedPayload = payload
+		statusMessages = append(statusMessages, message)
+	}
+	return statusMessages, nil
+}
+
+func (m *StatusMessage) HandleApplicationMetadata() error {
+	message, err := applicationmetadata.Handle(m.DecryptedPayload)
+	// Not an applicationmetadata message, calculate ID using the previous
+	// signature
+	if err != nil {
+		m.ID = MessageID(m.SigPubKey(), m.DecryptedPayload)
+		return nil
+	}
+
+	recoveredKey, err := message.RecoverKey()
+	if err != nil {
+		return err
+	}
+	m.ApplicationMetadataLayerSigPubKey = recoveredKey
+	m.DecryptedPayload = message.Payload
+	m.ID = MessageID(m.SigPubKey(), m.DecryptedPayload)
+	return nil
+
+}
+
+func (m *StatusMessage) HandleApplication() error {
+	originalPayload := m.DecryptedPayload
+	payload := make([]byte, len(originalPayload))
+	copy(payload, originalPayload)
+	// This modifies the payload
+	value, err := decodeTransitMessage(payload)
```

Comment:
How does this function modifies payload? We need to fix this if that's the case.

Reply:
`bytes.NewBuffer` modifies the the `[]byte` array passed in, I can copy it inside `decodeTransitMessage` so that `decodeTransitMessage` does not modify it, what do you think?

Reply:
```
NewBuffer creates and initializes a new Buffer using buf as its initial contents. The new Buffer takes ownership of buf, and the caller should not use buf after this call.
```

Right. Maybe we can do copying in `decodeTransitMessage`? I think it should be clear that whatever payload you pass to `decodeTransitMessage`, it should be unchanged.